### PR TITLE
Group and ungroup split transactions

### PR DESCRIPTION
**Pull request by joel-jeremy**
Easily turn multiple transactions into a split transaction or separate a split transaction into multiple individual ones.

# Make as split transaction
image

# Unsplit transaction
image

### CODE REVIEW

**Comment by psybers** [comment]
Right now if I want to pull *one* of the splits out, I have to separate the entire transaction, then join back the other splits together.

Is there a way, since individual splits can be selected, to just have it so you can select a subset of the splits and pull those out into separate transactions, leaving the rest together as a split?

**Comment by psybers** [comment]
Here are some bugs:

1. It seems to let me pick two transactions from different accounts and make them into a split.
2. When you make a new split from two transactions, the 'parent' transaction that is created does not have a payee.  Only the splits do.
3. I am not sure how the undo/redo system works but if you can somehow group several steps into a single "undo" operation, that might make sense.  Otherwise you have to ctrl-z twice to actually undo the join operation.

**Comment by psybers** [comment]
@joel-jeremy Instead of always making a "Split" payee, could it be a bit smarter?  If the transactions being grouped are all the same Payee, it should use that.  If not, I am not sure of the best fallback case, but perhaps we fall back to either the one that is most common or the newest?  I do not have any payee named "Split" and would not like having that, personally.

**Comment by psybers** [comment]
I am unsure why it renders differently.  Here is a real split (from the test data):
image

Here is one I created:
image

Notice the payees of the children are not greyed out like the first one.

**Comment by joel-jeremy** [comment]
@psybers Thank you for testing this out! Pushed some changes. Seems to fix most of the issues, and now it's going to allow pulling out one or more child transactions, not just as a whole.

**Comment by psybers** [comment]
Here's another edge case:

image

It gives the option to unsplit here.

**Comment by joel-jeremy** [comment]
@psybers The code handles that edge-case, it ignores the non-split transactions.

**Comment by psybers** [comment]
If I pick all but 1 of the splits and then pull them out, I am left with this parent with one child:

image

Probably it should just go to a normal, non-split transaction?

**Comment by psybers** [comment]
> @psybers The code handles that edge-case, it ignores the non-split transactions.

I was thinking it makes more sense to not show the menu option in this edge case.

**Comment by psybers** [comment]
General UX principles state the menu item should probably always be shown, and just grey out when it isn't available.  Similar to the "Make transfer" option.

The wording of the menu item could maybe be improved, but I am struggling with some good suggestions.

- Join transactions
- Merge into split
- Make split transaction

**Comment by joel-jeremy** [comment]
Maybe we should consider introducing a special "Split" payee just as we have a special "Transfer" category. I always found it a bit odd having to set a payee on the parent transaction when the child transactions can have differing payees that are unrelated. 

You didn't really pay for anything with the parent transaction, you already denoted that on it's child transactions. So more likely than not, the parent payee is just duplicating the info on the child transactions.

**Comment by Teprifer** [comment]
> Maybe we should consider introducing a special "Split" payee just as we have a special "Transfer" category. I always found it a bit odd having to set a payee on the parent transaction when the child transactions can have differing payees that are unrelated.
> 
> You didn't really pay for anything with the parent transaction, you already denoted that on it's child transactions. So more likely than not, the parent payee is just duplicating the info on the child transactions.

Maybe, but this logic doesn't quite follow when there is only one actual transaction and the split is to apportion parts to different categories - it would be annoying and superfluous having to set the same payee repeatedly vs the once you do now.

**Comment by joel-jeremy** [comment]
> Maybe, but this logic doesn't quite follow when there is only one actual transaction and the split is to apportion parts to different categories - it would be annoying and superfluous having to set the same payee repeatedly vs the once you do now.

To help with that, we can have the "Split" payee render as a button, which opens the payee autocomplete to set all child transactions to the same payee, but parent will still have the "Split" payee.

**Comment by psybers** [comment]
@joel-jeremy Can you provide a concrete use case of where it makes sense to take two separate transactions, from two different payees, and then join them together as a "split"?  I am struggling to envision this case.

I can clearly see the other case, where you might want to pull part of a split out (I would use that every week!).  An example is if you split a bill with your partner and want to pull their part out to track separately.  You could leave it as a split and just categorize it differently but it is nice to see it separate.  Then you could delete it.

**Comment by psybers** [comment]
Any ideas what this would/would not do if you are using a bank sync?  If I pull out a split, or join two things together will it suddenly try to sync the original transaction back in?

**Comment by joel-jeremy** [comment]
Example of how we could render the parent transaction payee:

image

When split button is clicked:
image\

Updates all the child transaction payees:
image

**Comment by joel-jeremy** [comment]
> Can you provide a concrete use case of where it makes sense to take two separate transactions, from two different payees, and then join them together as a "split"? I am struggling to envision this case.

@psybers  Personally, I use split transaction for transactions that have fees. I add the fee as a child transaction with a different payee e.g.

image

**Comment by joel-jeremy** [comment]
> Any ideas what this would/would not do if you are using a bank sync? If I pull out a split, or join two things together will it suddenly try to sync the original transaction back in?

Haven't done a thorough testing but I believe as long as the `imported_id` is still set, the sync logic  will match that regardless if the transaction became a child transaction or vice versa.

**Comment by psybers** [comment]
> > Can you provide a concrete use case of where it makes sense to take two separate transactions, from two different payees, and then join them together as a "split"? I am struggling to envision this case.
> 
> @psybers Personally, I use split transaction for transactions that have fees. I add the fee as a child transaction with a different payee e.g.
> 
> image

Interesting! And that fee is not tacked onto the credit card charge?  I guess in my life that is typically what happens, so there it is more like a normal split.

**Comment by psybers** [comment]
> Example of how we could render the parent transaction payee:

And here, the "Split" payee is a magic payee (kind of like Transfers)?  So it would need to be filtered out of the payees page (or listed and grey, like transfers)? Can you explicitly pick it, and if so will it need to be rendered differently in the list of payees (similar to transfers)?

**Comment by joel-jeremy** [comment]
> Interesting! And that fee is not tacked onto the credit card charge? I guess in my life that is typically what happens, so there it is more like a normal split.

The fee is not included in the CC statement. The mobile wallet app I use charges a fee when I pay my CC through it.

> And here, the "Split" payee is a magic payee (kind of like Transfers)? So it would need to be filtered out of the payees page (or listed and grey, like transfers)? Can you explicitly pick it, and if so will it need to be rendered differently in the list of payees (similar to transfers)?

Yes, it will be like transfers. The parent transaction will have no payee at all (on the database level). We will just provide the button as a means to easily set all the payee on all the child transactions.

**Comment by psybers** [comment]
@joel-jeremy Looks good!

Right now you have it showing the select payee as sort of a modal (but not a modal).  Is there a way to have it just drop down, like it normally would?  It also does not have the 'manage payees' button that would normally appear in that drop down.

**Comment by joel-jeremy** [comment]
@psybers I think the bulk edit modal would fit better here since it's going to be updating multiple payees (all children payees).

I wonder if we should add a db migration to remove all payees from parent transactions or just leave those that already have a payee as-is.

**Comment by psybers** [comment]
> I wonder if we should add a db migration to remove all payees from parent transactions or just leave those that already have a payee as-is. Would like to get you thoughts on this too @MatissJanis.

@joel-jeremy I assume it would only be removed if the parent's payee matched all of the children's payees?

**Comment by joel-jeremy** [comment]
> I assume it would only be removed if the parent's payee matched all of the children's payees?

@psybers It will remove all parent's payee regardless of the children's payees. When this PR gets merged, all parent transactions moving forward would have no payees. The special Split payee button would only set the child transaction's payee.

File: packages/desktop-client/e2e/transactions.test.js-snapshots/Transactions-filters-transactions-by-category-1-chromium-linux.png
Comment (psybers) [inline packages/desktop-client/e2e/transactions.test.js-snapshots/Transactions-filters-transactions-by-category-1-chromium-linux.png:1]: Why did these changes?

**Comment by MatissJanis** [comment]
> I wonder if we should add a db migration to remove all payees from parent transactions or just leave those that already have a payee as-is. Would like to get you thoughts on this too @MatissJanis.

The removal of a specific payee from parent transactions - I can see this ending badly. There are definitely some users of Actual that are using splits and are setting different payees for the children/parent transactions. It's a feature (for better of for worse) we currently support. If we were to remove it - it has the potential to break these workflows. And if it's done via migration - there would be no way to recover. That's scary..

Is there a good argument for removing the functionality of allowing to manually set payees? Or is it mostly to make the product more "pure"?

--

While I am reviewing this here: any way we could add the merge/split options someplace else instead in the transaction menu? I think we have reached a breaking point in that menu - it has way too many things in it.

Some ideas: perhaps if you select > 1 transaction - a new "merge" button appears that creates the split transaction? And if you select child transactions then a "unsplit" button appears?

**Comment by joel-jeremy** [comment]
> The removal of a specific payee from parent transactions - I can see this ending badly. There are definitely some users of Actual that are using splits and are setting different payees for the children/parent transactions. It's a feature (for better or for worse) we currently support. If we were to remove it - it has the potential to break these workflows. And if it's done via migration - there would be no way to recover. That's scary..

I can see that. It is indeed safer to keep the existing parent payees as is - we can always ignore the parent payees in the UI and backend.  

> Is there a good argument for removing the functionality of allowing to manually set payees? Or is it mostly to make the product more "pure"?

My thoughts: https://github.com/actualbudget/actual/pull/2805#issuecomment-2138675861

> While I am reviewing this here: any way we could add the merge/split options someplace else instead in the transaction menu? I think we have reached a breaking point in that menu - it has way too many things in it.

> Some ideas: perhaps if you select > 1 transaction - a new "merge" button appears that creates the split transaction? And if you select child transactions then a "unsplit" button appears?

Agree that the menu is getting long. Though I think we should still keep the merge/split options in the selected transactions menu to keep things cohesive. Maybe we can make the "Edit Field" options a sub-menu which shows on hover?

**Comment by MatissJanis** [comment]
> Agree that the menu is getting long. Though I think we should still keep the merge/split options in the selected transactions menu to keep things cohesive. Maybe we can make the "Edit Field" options a sub-menu which shows on hover?

Personally - I'm not really a fan of sub-menus (i.e. deep nested menus). What about splitting it out into a separate button? Besides the "x transactions" - a "edit fields" button that opens the edit options.

Alternatively: we could hide the unavailable options from the menu. And actually.. I think we should hide them either way. That's part of the actual design philosophy - progressive discovery of features (i.e. a feature becomes visible only if it is available; otherwise - it's hidden).

**Comment by psybers** [comment]
> Alternatively: we could hide the unavailable options from the menu.

The only problem with hiding menu items is that it makes it difficult to get used to where things are in the menu.  Every time you open the menu, the item you want to click on can move around so you can't quickly move your mouse to what you are expecting because it might be in various places.  It forces you to read the whole menu every time you open it.

**Comment by MatissJanis** [comment]
> The only problem with hiding menu items is that it makes it difficult to get used to where things are in the menu. Every time you open the menu, the item you want to click on can move around so you can't quickly move your mouse to what you are expecting because it might be in various places. It forces you to read the whole menu every time you open it.

Muscle memory is a good argument to have. Though my assumption would be that the merge/un-merge button is always there. It's just the label and the functionality of it that is dynamic.

This wouldn't allow us to cover all edge-cases though (i.e. selecting multiple parents + children and then "merging" or "splitting" these via 2x individual buttons as there would be only one visible at all times), but for the sake of reducing the visual complexity - I think it's a fair compromise to make. And in such scenario - at worst it would mean 2x clicks on the same button.

File: packages/desktop-client/src/components/transactions/SelectedTransactionsButton.jsx [hunk 1]
```diff
@@ -18,83 +20,129 @@ export function SelectedTransactionsButton({
...
-  }, [selectedItems]);
+  }, [selectedIds]);
 
   const ambiguousDuplication = useMemo(() => {
-    const transactions = [...selectedItems].map(id => getTransaction(id));
+    const transactions = selectedIds.map(id => getTransaction(id));
 
     return transactions.some(t => t && t.is_child);
-  }, [selectedItems]);
+  }, [selectedIds, getTransaction]);
```
Comment (MatissJanis) [inline packages/desktop-client/src/components/transactions/SelectedTransactionsButton.jsx:43]: **👏 praise:** ‏good catch!

**Review by MatissJanis — APPROVED** [review]

**Comment by Zybrion** [comment]
Hello @joel-jeremy,
I'd like to provide some feedback regarding the split transaction entry.
I have always used splits like a `receipt` mechanism. The parent transaction has the full amount and each item bought (for example in a store or in a restaurant) gets its own split entry underneath the parent entry.

The issue I am now facing after this change is that about half of my transactions now have `split` as the `payee` which is really annoying because with a quick glance I cannot see from where / what I bought there. I'd have to manually open each and every `split transaction` in order to gain knowledge from where / what I bought.

Is it possible to integrate the suggestion from @psybers mentioned here? https://github.com/actualbudget/actual/pull/2805#issuecomment-2137879364
I heavily agree that when all child transactions have one common payee that this one payee should be shown when the split transaction is collapsed. If there are multiple payees in the child transaction, then I personally would be fine having `split` shown as a payee (although I'd still prefer to control it myself).

I have attached a screenshot to maybe better visualize my problem:
split_transactions

Thank you for your work!

**Comment by joel-jeremy** [comment]
Hey there 👋 thank you for your feedback! There is an ongoing PR for that: https://github.com/actualbudget/actual/pull/3049

**Comment by Zybrion** [comment]
Oh, very awesome! I am sorry I didn't see that one. Thank you! That looks very nice

**Comment by warstellar** [comment]
I am not sure what I'm doing wrong, but it doesn't seem to make any sense as to why I can see or not see this option:
image
image

**Comment by vcar65** [comment]
I can't see the options 'Mark as split transaction' nor 'Unspilt transactions'.
I'm running version v24.10.1 on Pikapods

**Pull request merged**